pkg/schema/error.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package schema
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/k14s/ytt/pkg/filepos"
+	"github.com/k14s/ytt/pkg/yamlmeta"
+)
+
+type errorWithContent struct {
+	contents string
+	Position *filepos.Position
+}
+
+func (e errorWithContent) Error() string {
+	panic("do not use directly")
+}
+
+func (e *errorWithContent) SetContext(doc *yamlmeta.Document) error {
+	contents, err := doc.RawDataAtLine(e.Position)
+
+	if err == nil {
+		e.contents = contents
+	}
+
+	return err
+}
+
+func NewInvalidError(found, expected, hint string, position *filepos.Position) error {
+	return &invalidError{
+		errorWithContent: errorWithContent{
+			Position: position,
+		},
+		Found:    found,
+		Expected: expected,
+		Hint:     hint,
+	}
+}
+
+type invalidError struct {
+	errorWithContent
+	Found    string
+	Expected string
+	Message  string
+	Hint     string
+}
+
+func (i invalidError) Error() string {
+	var result string
+	position := i.Position.AsCompactString()
+	lineContent := strings.TrimSpace(i.contents)
+	colonPosition := strings.Index(lineContent, ":")
```

Comment:
are we assuming that this would always be map key pair? what if key includes a ":"? is there a higher level to work with here instead of working at the string level?

Reply:
There are a couple of places in here that we're introducing an assumption that `MapItem.Key`'s are strings. We need to revisit and generalize.

Reply:
Errors now receive the entire node, no need to worry about whether there was a colon in the key or not. We also stopped trying to point to the value with carets so determining the colon position is no longer required.